### ISSUE TITLE
uncaught FILE_ENDED error

### ISSUE THREAD

**Issue by rhodgkins** [ZJONSSON/node-unzipper#75]
I think this is the same as #48 - I've opened a new issue as that was closed and changes have been made since then...!

After finding this out when moving over to the Open APIs (#74) I've been looking into it and found that it's caused by the following line:

https://github.com/ZJONSSON/node-unzipper/blob/835c041c8c3e43029f1a665c61b7f307f4d548dd/lib/PullStream.js#L66

Is this supposed to be emitting on `p` - the returned PassThrough stream instead?

I've been playing around with a fix [here](https://github.com/bookcreator/node-unzipper/tree/fix/FILE_ENDED/lib) which instead emits the error to the returned stream instead.
I've also had to include a similar fix as #57 (nulling out `self.cb`) and these changes fix the uncaught error and its propagated up to the Promise retuned from `unzipper.Open.file` but it breaks the `parseOneEntry.js: error - file ended` test case and I'm not sure why :-/

I hope this helps you debug it a bit! It might be the case that after the bug in the fix for #57 is implemented (#71) everything might work without this anyway...

**Comment by ZJONSSON** [comment]
I think this should be fixed now - please reopen if issue persists

**Comment by lucidkodo** [comment]
Hey guys! i'm not sure if this issue is back? 

But i was using `zipper.Extract()` in a `for` loop
```
for (let i = 0; i < files.length; i++) {
  // console.log(files[i]) // /home/angie/exporter/auto/unsubscribe/zip/something.csv.zip

  fs.createReadStream(files[i])
    .pipe(zipper.Extract({
      path: 'path/to/dest/'
    }))
    .on('error', e => {
      console.log(e)
    })
    .on('finish', () => {
      console.log('finished')
    })
}
```

and gave gave me this error: 
```
Error: FILE_ENDED
    at Parse.pull (/home/angie/exporter/node_modules/unzipper/lib/PullStream.js:80:28)
    at Parse.emit (events.js:189:13)
    at Parse.<anonymous> (/home/angie/exporter/node_modules/unzipper/lib/PullStream.js:20:10)
    at Parse.emit (events.js:194:15)
    at finishMaybe (_stream_writable.js:641:14)
    at endWritable (_stream_writable.js:649:3)
    at Parse.Writable.end (_stream_writable.js:589:5)
    at DuplexWrapper.<anonymous> (/home/angie/exporter/node_modules/duplexer2/index.js:29:14)
    at Object.onceWrapper (events.js:277:13)
    at DuplexWrapper.emit (events.js:194:15)
    at finishMaybe (/home/angie/exporter/node_modules/duplexer2/node_modules/readable-stream/lib/_stream_writable.js:630:14)
    at endWritable (/home/angie/exporter/node_modules/duplexer2/node_modules/readable-stream/lib/_stream_writable.js:638:3)
    at DuplexWrapper.Writable.end (/home/angie/exporter/node_modules/duplexer2/node_modules/readable-stream/lib/_stream_writable.js:594:41)
    at ReadStream.onend (_stream_readable.js:651:10)
    at Object.onceWrapper (events.js:277:13)
    at ReadStream.emit (events.js:194:15)
    at endReadableNT (_stream_readable.js:1125:12)
    at process._tickCallback (internal/process/next_tick.js:63:19)
finished
finished
finished
Done in 5.99s.
```

Any help would be much appreciated! Thanks for the package! :)

**Comment by ZJONSSON** [comment]
`FILE_ENDED` is an error that is thrown if the zip file is not valid, i.e. the reader reads past of the end of the zip file without concluding the parse.  Is it possible that one of the files is broken?

You might want to try async await and do one file at a time and see if it makes a difference, or at the very least spot the offending file.
```js
async function extract() {
  for (let i = 0; i < files.length; i++) {
    console.log(files[i]);
    await fs.createReadStream(files[i])
      .pipe(zipper.Extract({path: 'path/to/dest'}))
      .on('error',console.log)
      .on('finish', () => console.log('finised')
      .promise()
  }
}
```

Or you could use the `Open` method (which is technically better as it uses the Central Directory for offset information):

```js
async function extract() {
  for (let i = 0; i < files.length; i++) {
    console.log(files[i]);
    const directory = await zipper.Open.file(files[i]);
    await directory.extract({'path/to/dest'})
  }
}
```

**Comment by lucidkodo** [comment]
@ZJONSSON thank you so much for your prompt reply! 
There was an empty csv file in my array!

I'm using the first suggested approach as i need to do some extra processing once the file is done unzipping. thanks again!

**Comment by jlommori** [comment]
@ZJONSSON I'm experiencing the same file_ended issues when using the `zipper.Open.file(file)` method. I'm looping through paths for 10 zip files, all of which are valid (at least can be opened by the OS and have contents, both dirs and files). I receive the error inconsistently each time I run the code. Sometimes its all files, sometimes only some of them. It has me stumped. Any suggestions for troubleshooting?

**Comment by ZJONSSON** [comment]
Can you confirm you are using the latest version of `unzipper` and also share the code?   If the zip files are not sensitive and you want me to replicate the issue you can send them to `ziggy.jonsson.nyc@gmail.com`

**Comment by jlommori** [comment]
@ZJONSSON Of course as I extract the code to post here and test it, everything works when the section isolated from everything else. So I guess I have some additional troubleshooting to do within the whole program. Thank you for the willingness to assist!
